### PR TITLE
Use NSE and {dplyr}

### PR DESCRIPTION
**Pull request by kellijohnson-NOAA**
NSE functions for working with biological data that will in theory work on data from any group

## Features
* `bin()`
* `quantify()`

## To do
- [ ] tests
- [ ] `format_*()` wrappers
- [ ] vignette
- [ ] more examples for `quantify()`

### CODE REVIEW

**Comment by chantelwetzel-noaa** [comment]
I don't have time to go over all the examples, but the ones I looked at for add_bin() and bin() look good. The only comment I have so far is that the add_bin() function is returning a tibble.  I would suggest modifying the returned object to a data frame in the function.

**Comment by iantaylor-NOAA** [comment]
TL;DR: there are a variety of reasons why we might start our data or population length bins above zero in Stock Synthesis, so I think the bin() function should facilitate that, even if starting at 0 makes sense most of the time.

For standard length comp data in Stock Synthesis, the expected value for the first data bin includes all expected observations below the lower bound of that bin (from User Manual [here](https://nmfs-stock-synthesis.github.io/doc/SS330_User_Manual.html#:~:text=When%20processing%20data%20to%20be%20input%20into%20SS3%2C%20all%20observed%20fish%20of%20sizes%20smaller%20than%20the%20first%20bin%20should%20be%20added%20to%20the%20first%20bin%20and%20all%20observed%20fish%20larger%20than%20the%20last%20bin%20should%20be%20condensed%20into%20the%20last%20bin.)): 
> When processing data to be input into SS3, all observed fish of sizes smaller than the first bin should be added to the first bin and all observed fish larger than the last bin should be condensed into the last bin.

So in that case, we should always be binning the data starting at 0 as you suggest even if the bins we enter to the model start at a higher value.

However, for generalized size comp data you have to choose whether to aggregate below that first bin or not (from [here](https://nmfs-stock-synthesis.github.io/doc/SS330_User_Manual.html#:~:text=There%20are%20two%20options,the%20first%20size%20bin.)):
> There are two options for treating fish that in population size bins are smaller than the smallest size frequency bin.
> 
> Option 1: By default, these fish are excluded (unlike length composition data where the small fish are automatically accumulated up into the first bin.)
> 
> Option 2: If the first size bin is given as a negative value, then accumulation is turned on and the absolute value of the entered value is used as the lower edge of the first size bin.
> 
So in that case, users may want to make sure they aren't including fish below their minimum bin (though they should still get a warning if there are observations below that bin).

As for population bins, I don't think you can have the lower bound of the first data length bin below the first population length bin and I think the goal of starting the population bins above 0 is to just avoid modeling a range of lengths that don't matter for the dynamics. In some cases, like spiny dogfish, size at birth is > 20 cm, so smaller fish simply don't exist. In other cases, there's a larval stage and the age-0 fish are simply never observed, so folks set the Amin (or A1) value above 0 which causes a linear ramp in growth for smaller fish as described in the User Manual [here](https://nmfs-stock-synthesis.github.io/doc/SS330_User_Manual.html#:~:text=At%20the%20time%20of%20settlement,been%20retained%20in%20the%20model.):

> At the time of settlement, fish are assigned a size equal to the lower edge of the first population size bin and they grow linearly until they reach the age A1. A warning is generated if the first population length bin is greater than 10 cm as this seems an unreasonably large value for a larval fish. A1 is in terms of real age elapsed since birth. All fish advance to the next integer age on January 1, regardless of birth season. For example, consider a 2 season model with some recruitment in each season and with each season’s recruits coming from the same GP. At the end of the first year, the early born fish will be larger but both of the seasonal cohorts will advance to an integer age of 1 on Jan 1 of the next year. The full growth curve is still calculated below A1, but the size-at-age used is the linear replacement. Because the linear growth trajectory can never go negative, there is no need for the additive constant to the standard deviation (necessary for the growth model used in SS2 V1.x), but the option to add a constant has been retained in the model.

Since the starting point of that linear ramp is based on the population length bins, the choice of the first bin impacts the growth curve (illustrated in the fig below), which can be distracting if the linear ramp is too steep or too shallow even if those fish are playing no role in the dynamics or observations.

image

File: R/bin.R [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+#'
+#' # add_bin() works like dplyr::mutate() instead of dplyr::summarise()
+#' # (result has same number of rows as input)
+#' original <- add_bin(
+#'   data = bio_nwfsc_combo %>% group_by(Year, Sex = codify_sex(Sex)),
+#'   var = Length_cm,
+#'   bins = seq(0, 300, by = 20)
+#' )
+#' # The following have different counts per bin because bins in original
+#' # are closed on the right and open on the left and when we use bespoke
```
Comment (chantelwetzel-noaa) [inline R/bin.R:105]: Is this correct? The original example on lines 99-103 results in closed bins on the left and open on the right (e.g., [0, 20)).

File: R/bin.R [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+#' # filled in and sum of counts should equal the number of rows of data
+#' dplyr::count(summary, Length_cm_bin)
+#' sum(summary[["Length_cm_n"]]) == NROW(bio_nwfsc_combo)
+#'
+#' # add_bin() works like dplyr::mutate() instead of dplyr::summarise()
+#' # (result has same number of rows as input)
+#' original <- add_bin(
+#'   data = bio_nwfsc_combo %>% group_by(Year, Sex = codify_sex(Sex)),
+#'   var = Length_cm,
+#'   bins = seq(0, 300, by = 20)
```
Comment (chantelwetzel-noaa) [inline R/bin.R:102]: I am a little concerned about the departure in this binning approach compared to the current version in the package. In the current package if a user provided bins = seq(10, 60, 2) the user would get length compositions with columns for [0-10), [10-12),..., [60, Inf). In this new approach if a user provided these same bins, they would get a warning about fish less than 10 cm and greater than 60 with those fish assigned a bin of NA requiring the user to adjust the provided bins = c(0, seq(10, 60, 2), Inf) to get the same functionality.  This is okay but I think we should consider if we want to automatically add bins below or above those provided if the full length rang is not included.

File: R/bin.R (hunk 1 above)
Comment (kellijohnson-NOAA) [reply R/bin.R:105]: Thanks @chantelwetzel-noaa I have since changed this to 
```
#' # The following have different counts per bin because bins in original
#' # are closed on the left and open on the right and when we use bespoke
#' # dplyr code here, the bins are open on left.
```

File: R/bin.R (hunk 2 above)
Comment (kellijohnson-NOAA) [reply R/bin.R:102]: We can write any wrapper function to accommodate adding bins to the tails of the distribution but I wanted the base functions to not have any hidden characteristics/methods. That is, you get what you put in otherwise it gets too messy to use the same base code for multiple data sets / legacy workups.

Comment (iantaylor-NOAA) [reply R/bin.R:102]: I like the "get what you put in" approach and I think it's fine to make users take an extra step to pull in samples that are below their lowest bin boundary. But I think the warning should only occur if samples are outside the bin range. If the user chooses bins = seq(10,60,2) and that includes all the samples, I don't think they should get a warning about their lower bound not being at zero.

Comment (kellijohnson-NOAA) [reply R/bin.R:102]: Thanks @iantaylor-NOAA I have since changed the warning to a message that checks to see if there are lower or higher values in the binning groups.

Comment (chantelwetzel-noaa) [reply R/bin.R:102]: @iantaylor-NOAA in the example you gave that if a user specifies bins = seq(10, 60, 2), it is my understanding that this input to the bin functions would fail to give the user the length-plus group on fish of 60+.  This is heart of my concern that using the "old" input would fail to produce what the user was expecting to receive.  I fully support the idea that users should "get what you put in" approach, but perhaps there should be some checks, warnings, and/or corrections within the function when samples are being excluded due to the function input, especially if we opt to change the input logic.  I am fully open to ideas here.

**Comment by iantaylor-NOAA** [comment]
I like the @kellijohnson-NOAA proposal for a wrapper function, e.g. `bin_for_ss3()` (only with a better name) where the input `seq(10, 60, 2)` would automatically get converted to `c(0, seq(10, 60, 2), Inf)`, and then the [0, 10) and [10, 12) bins would get combined, perhaps with a message like "3 out of the 15 fish in the 10-12cm bin were less than 10 cm". And the [58, 60) bin would be grouped with the [60, Inf) with a similar message about how many fish were over 60cm.

**Comment by chantelwetzel-noaa** [comment]
I think we are approach some level of consensus on these proposed functions, @kellijohnson-NOAA are there additional components of the proposed that you would like to receive feedback on?

**Comment by chantelwetzel-noaa** [comment]
@kellijohnson-NOAA Did you want to create a new pull request for this still?

**Comment by kellijohnson-NOAA** [comment]
I will close it and open a new one when ready. Thanks for the reminder @chantelwetzel-noaa 😍
